### DISCUSSION TITLE
Trouble overwriting pg.PlotWidget.addItem

### DISCUSSION THREAD

**Discussion by Seaspider1** [pyqtgraph/pyqtgraph#2820] [Q&A]
I am designing an application in which plots are created and updated dynamically. Specifically, I need attributes that are contained with the data to be able to update the plot features (labels, scaling, etc.). When I attempted to subclass pg.PlotWidget, I was unable to get the addItem method to be called successfully using traditional subclassing. Here is some code demonstrating what I've done and how I got around it. Any ideas on why addItem is not actually a method of PlotWidget, but a PlotItem? I couldn't find a path between the two in the source code.

	class MyCurve(pg.PlotCurveItem):
		def __init__(self, *args, x_label='', y_label='', **kwargs):
			super().__init__(*args, **kwargs)
			self.x_label = x_label
			self.y_label = y_label

	class MyPlotWidget(pg.PlotWidget):
		def __init__(self):
			super().__init__()

		def addItem(self, item, *args):  # This never gets called, instead pg.graphicsItems.PlotItem.PlotItem.PlotItem.addItem() is called
			if hasattr(item, 'x_label'):
				self.setLabel('bottom', item.x_label)
			if hasattr(item, 'y_label'):
				self.setLabel('left', item.y_label)
			super().additem(item, *args)
			
			
	# Function call in code:

	# ...

	# Define a curve with label attributes
	curve = MyCurve([0, 1, 4, 9, 16], x_label='Time', y_label='Amplitude')

	my_plot = MyPlotWidget()
	my_plot.addItem(curve)  # Plot should modify labels when data is added (but doesn't)

	# ...

	# Inspection of my_plot shows <__main__.MyPlotWidget(0x1b9a9e7e9d0) at 0x000001B9AB47D280>
	# Inspection of my_plot.addItem shows <bound method PlotItem.addItem of <bound method PlotItem.addItem of <pyqtgraph.graphicsItems.PlotItem.PlotItem.PlotItem(0x1b9a9e8e690, pos=0,0, flags=(ItemUsesExtendedStyleOption|ItemSendsGeometryChanges)) at 0x000001B9AB47C080>>

	# Hacky definition that works correctly
	class MyPlotWidget(pg.PlotWidget):
		def __init__(self):
			super().__init__()

			self.old_additem = self.addItem

			def addItem(self, item, *args):
				if hasattr(item, 'x_label'):
					self.setLabel('bottom', item.x_label)
				if hasattr(item, 'y_label'):
					self.setLabel('left', item.y_label)
				self.old_additem(item, *args)

			self.addItem = lambda item, *args: addItem(self, item, *args)

**Comment by pijyoi** [comment]
You could do `self.getPlotItem().addItem(...)`
